vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -211,6 +212,160 @@ void learn_or_predict(ccb& data, multi_learner& base, multi_ex& examples)
   examples[0]->pred.decision_scores = data.decision_scores;
 }
 
+void print_decision_scores(int f, decision_scores_t& decision_scores)
+{
+  if (f >= 0)
+  {
+    std::stringstream ss;
+    for (auto decision : decision_scores)
+    {
+      std::string delimiter = "";
+      for (auto action_score : decision)
+      {
+        ss << delimiter << action_score.action << ':' << action_score.score;
+        delimiter = ",";
+      }
+      ss << '\n';
+    }
+    ss << '\n';
+    ssize_t len = ss.str().size();
+    ssize_t t = io_buf::write_file_or_socket(f, ss.str().c_str(), (unsigned int)len);
+    if (t != len)
+      std::cerr << "write error: " << strerror(errno) << std::endl;
+  }
+}
+
+void print_update(vw& all, std::vector<example*> decisions, decision_scores_t decision_scores, size_t num_features)
+{
+  if (all.sd->weighted_examples() >= all.sd->dump_interval && !all.quiet && !all.bfgs)
+  {
+    std::string label_str = "";
+    std::string delim = "";
+    int counter = 0;
+    for (auto decision : decisions)
+    {
+      counter++;
+
+      auto outcome = decision->l.conditional_contextual_bandit.outcome;
+      if (outcome == nullptr)
+      {
+        label_str += delim;
+        label_str += "?";
+      }
+      else
+      {
+        label_str += delim;
+        label_str += std::to_string(outcome->probabilities[0].action);
+        label_str += ":";
+        label_str += std::to_string(outcome->cost);
+      }
+
+      delim = ",";
+
+      // Stop after 2...
+      if (counter > 1 && decisions.size() > 2)
+      {
+        label_str += delim;
+        label_str += "...";
+        break;
+      }
+    }
+    std::ostringstream label_buf;
+    label_buf << std::setw(all.sd->col_current_label) << std::right << std::setfill(' ') << label_str;
+
+
+    std::string pred_str = "";
+    delim = "";
+    counter = 0;
+    for (auto decision : decision_scores)
+    {
+      counter++;
+      pred_str += delim;
+      pred_str += std::to_string(decision[0].action);
+      delim = ",";
+
+      // Stop after 3...
+      if (counter > 2)
+      {
+        pred_str += delim;
+        pred_str += "...";
+        break;
+      }
+    }
+    std::ostringstream pred_buf;
+    pred_buf << std::setw(all.sd->col_current_predict) << std::right << std::setfill(' ') << pred_str;
+
+
+    all.sd->print_update(all.holdout_set_off, all.current_pass, label_buf.str(), pred_buf.str(), num_features,
+        all.progress_add, all.progress_arg);
+  }
+}
+
+inline float get_unbiased_cost(CCB::conditional_contexual_bandit_outcome* outcome, uint32_t action, float offset = 0.)
```

Comment:
Why can't we share this function with other implementations?

Reply:
CCB stores the cost and probability in it's own outcome structure and so we need a function that understands this. I can definitely move it to a common location though

Reply:
So the difference between CCB class and CB class is that CCB _can_ contain the entire pdf for that slot. So far that information has not been utilized on the reduction, so do we actually need it? If we don't then CCB can simply use the CB class

Reply:
@JohnLangford can you comment on the above? ^